spmd/tensor/device_mesh.py

```diff
@@ -274,41 +275,69 @@ def scatter(
-        # makes sure communication result is properly waited before subsequent
-        # read operations.
-        to_scatter = [CommTensor(tensor.contiguous()) for tensor in scatter_list]
         dim_group = self._dim_groups[mesh_dim]
         # src need to be global rank
         src_for_dim = 0
         if dim_group is not GroupMember.WORLD:
             src_for_dim = get_global_rank(dim_group, 0)
 
-        # N.B.: the `tensor` below will be a CommTensor too due to CommTensor's
-        # propagation rule: propagte wrapping until communication is called.
-        # This is necessary in order to properly trigger CommTensor's dispatch
-        # function for scatter_.
-        tensor = torch.empty_like(to_scatter[my_coordinate])
-        if src_for_dim == get_rank():
-            scatter(
-                tensor,
-                scatter_list=to_scatter,
-                src=src_for_dim,
-                group=dim_group,
-            )
+        scatter_list = list(
+            tensor_to_scatter.tensor_split(num_chunks, dim=tensor_dim)
+        )
+        if self._backend == "gloo":
+            # gloo does not support uneven scattering, we need to resize tensor
+            # before really calling scatter, the first shape is always bigger
+            # TODO: remove this logic once ProcessGroupGloo suupport uneven list
+            scatter_shape = scatter_list[0].shape
+            my_rank_shape = scatter_list[my_coordinate].shape
+            to_scatter = []
+            for scatter_tensor in scatter_list:
+                scatter_tensor = scatter_tensor.contiguous()
+                if scatter_tensor.shape != scatter_shape:
+                    scatter_tensor.resize_(scatter_shape)
+                to_scatter.append(scatter_tensor)
+
+            # N.B.: the `tensor` below will be a CommTensor too due to CommTensor's
+            # propagation rule: propagte wrapping until communication is called.
+            # This is necessary in order to properly trigger CommTensor's dispatch
+            # function for scatter_.
+            tensor = torch.empty_like(to_scatter[my_coordinate])
+            if src_for_dim == get_rank():
+                scatter(
+                    tensor,
+                    scatter_list=to_scatter,
+                    src=src_for_dim,
+                    group=dim_group,
+                )
+            else:
+                scatter(
+                    tensor, scatter_list=None, src=src_for_dim, group=dim_group
+                )
+
+            # resize to uneven size if needed
+            if my_rank_shape != tensor.shape:
+                tensor.resize_(my_rank_shape)
+            return tensor
+        elif self._backend == "nccl":
+            to_scatter = [tensor.contiguous() for tensor in scatter_list]
+
+            tensor = torch.empty_like(to_scatter[my_coordinate])
+            if src_for_dim == get_rank():
+                scatter(
```

Comment:
Does this mean NCCL actually supports uneven scatter? But I remembered we did resize_ for ShardedTensor. For learning purpose, why is it working now?

Reply:
pg nccl recently supported this I think, although gloo not, and likely uneven scatter is not performing better than even one, so I updated to do pad/unpad before/after so that it could reuse the even size collective.